File: common/src/main/java/af/shizuku/common/util/EnvironmentUtils.java
package af.shizuku.common.util;

import android.app.UiModeManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.os.Build;
import android.os.SystemProperties;

public class EnvironmentUtils {

    public static boolean isWatch(Context context) {
        return (context.getSystemService(UiModeManager.class).getCurrentModeType()
                == Configuration.UI_MODE_TYPE_WATCH);
    }

    public static boolean isTelevision(Context context) {
        return (context.getSystemService(UiModeManager.class).getCurrentModeType()
                == Configuration.UI_MODE_TYPE_TELEVISION ||
                context.getPackageManager().hasSystemFeature(PackageManager.FEATURE_LEANBACK));
    }

    public static boolean isTlsSupported(Context context) {
        if (isTelevision(context))
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU;
        else 
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.R;
    }

    public static int getFullSdkVersion() {
        if (Build.VERSION.SDK_INT >= 36) {
            try {
                java.lang.reflect.Field field = Build.VERSION.class.getField("SDK_INT_FULL");
                return field.getInt(null);
            } catch (Exception e) {
                return Build.VERSION.SDK_INT * 100;
            }
        } else {
            return Build.VERSION.SDK_INT * 100;
        }
    }

    public static boolean isSamsung() {
        return Build.MANUFACTURER.equalsIgnoreCase("samsung");
    }

    public static int getOneUiVersion() {
        if (!isSamsung()) return -1;
        try {
            java.lang.reflect.Field field = Build.VERSION.class.getDeclaredField("SEM_PLATFORM_INT");
            return ((int) field.get(null) - 90000) / 10000;
        } catch (Exception e) {
            if (Build.VERSION.SDK_INT >= 36) return 8;
            return -1;
        }
    }

    public static boolean isOneUi8() {
        return isSamsung() && getOneUiVersion() >= 8;
    }

    public static boolean isOppo() {
        return Build.MANUFACTURER.equalsIgnoreCase("oppo") || Build.MANUFACTURER.equalsIgnoreCase("realme");
    }

    public static boolean isOnePlus() {
        return Build.MANUFACTURER.equalsIgnoreCase("oneplus");
    }

    public static boolean isXiaomi() {
        return Build.MANUFACTURER.equalsIgnoreCase("xiaomi") || Build.MANUFACTURER.equalsIgnoreCase("redmi") || Build.MANUFACTURER.equalsIgnoreCase("poco");
    }

    public static boolean isTCL() {
        return Build.MANUFACTURER.equalsIgnoreCase("tcl");
    }

    public static boolean isDeX(Context context) {
        Configuration config = context.getResources().getConfiguration();
        try {
            java.lang.reflect.Field field = config.getClass().getField("semDesktopModeEnabled");
            return field.getInt(config) == 1;
        } catch (Exception e) {
            return (config.uiMode & Configuration.UI_MODE_TYPE_MASK) == Configuration.UI_MODE_TYPE_DESK;
        }
    }

    public static boolean isSecondaryUser() {
        try {
            return !android.os.Process.myUserHandle().toString().contains("{0}");
        } catch (Exception e) {
            return false;
        }
    }

    public static int getAdbTcpPort() {
        int port = SystemProperties.getInt("service.adb.tcp.port", -1);
        if (port == -1) port = SystemProperties.getInt("persist.adb.tcp.port", -1);
        return port;
    }
}

File: manager/src/main/java/af/shizuku/manager/utils/EnvironmentUtils.kt
package af.shizuku.manager.utils

import android.app.UiModeManager
import android.content.Context
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.os.Build
import android.os.SystemProperties
import af.shizuku.manager.ShizukuApplication
import af.shizuku.manager.ShizukuSettings
import com.topjohnwu.superuser.Shell

private val appContext: Context
    get() = try {
        ShizukuApplication.appContext
    } catch (e: UninitializedPropertyAccessException) {
        // Fallback for very early access if possible, or just rethrow with better message
        throw IllegalStateException("EnvironmentUtils.appContext accessed before ShizukuApplication.appContext was initialized", e)
    }

object EnvironmentUtils {

    @JvmStatic
    fun isWatch(): Boolean = af.shizuku.common.util.EnvironmentUtils.isWatch(appContext)

    @JvmStatic
    fun isTelevision(): Boolean = af.shizuku.common.util.EnvironmentUtils.isTelevision(appContext)

    fun isTlsSupported(): Boolean = af.shizuku.common.util.EnvironmentUtils.isTlsSupported(appContext)

    fun isWifiRequired(): Boolean {
        return (getAdbTcpPort() <= 0 || !ShizukuSettings.getTcpMode())
    }

    private var isRootedCached: Boolean? = null

    fun isRooted(): Boolean {
        if (isRootedCached == null) {
            isRootedCached = checkSuExists()
        }
        return isRootedCached!!
    }

    private fun checkSuExists(): Boolean {
        val paths = System.getenv("PATH")?.split(":") ?: return false
        for (path in paths) {
            if (java.io.File(path, "su").exists()) return true
        }
        return false
    }

    @JvmStatic
    fun getFullSdkVersion(): Int = af.shizuku.common.util.EnvironmentUtils.getFullSdkVersion()

    @JvmStatic
    fun isSamsung(): Boolean = af.shizuku.common.util.EnvironmentUtils.isSamsung()

    @JvmStatic
    fun getOneUiVersion(): Int = af.shizuku.common.util.EnvironmentUtils.getOneUiVersion()

    @JvmStatic
    fun isOneUi8(): Boolean = af.shizuku.common.util.EnvironmentUtils.isOneUi8()

    @JvmStatic
    fun isOppo(): Boolean = af.shizuku.common.util.EnvironmentUtils.isOppo()

    @JvmStatic
    fun isOnePlus(): Boolean = af.shizuku.common.util.EnvironmentUtils.isOnePlus()

    @JvmStatic
    fun getColorOsVersion(): String {
        return SystemProperties.get("ro.build.version.opporom", "unknown")
    }

    @JvmStatic
    fun isXiaomi(): Boolean = af.shizuku.common.util.EnvironmentUtils.isXiaomi()

    @JvmStatic
    fun getHyperOsVersion(): String {
        return SystemProperties.get("ro.miui.ui.version.name", "unknown")
    }

    @JvmStatic
    fun isTCL(): Boolean = af.shizuku.common.util.EnvironmentUtils.isTCL()

    @JvmStatic
    fun isDeX(context: Context): Boolean = af.shizuku.common.util.EnvironmentUtils.isDeX(context)

    @JvmStatic
    fun isSecondaryUser(): Boolean = af.shizuku.common.util.EnvironmentUtils.isSecondaryUser()

    fun getAdbTcpPort(): Int {
        var port = af.shizuku.common.util.EnvironmentUtils.getAdbTcpPort()
        if (port == -1 && isTelevision() && !isTlsSupported()) port = ShizukuSettings.getTcpPort()
        return port
    }

    /**
     * Resolves a SAF directory URI to a physical file path for a given filename.
     */
    @JvmStatic
    fun resolveExportedPath(filename: String): String? {
        val uriStr = ShizukuSettings.getExportDirUri() ?: return null
        return try {
            val uri = android.net.Uri.parse(uriStr)
            val docId = android.provider.DocumentsContract.getTreeDocumentId(uri)

            // Check for common volume patterns
            val basePath = when {
                docId.startsWith("primary:") -> {
                    val relative = docId.removePrefix("primary:")
                    if (relative.isEmpty()) "/storage/emulated/0"
                    else "/storage/emulated/0/$relative"
                }
                docId.contains(":") -> {
                    val parts = docId.split(":")
                    val volumeId = parts[0]
                    val relative = parts[1]
                    "/storage/$volumeId/$relative"
                }
                docId.startsWith("Download") || docId.startsWith("Documents") || docId.startsWith("Movies") -> {
                    "/storage/emulated/0/$docId"
                }
                else -> {
                    if (!docId.startsWith("/")) "/storage/emulated/0/$docId"
                    else docId
                }
            }

            val path = basePath.replace("//", "/") + "/" + filename
            path.replace("//", "/")
        } catch (e: Exception) {
            null
        }
    }
}
